elyra/pipeline/component.py

```diff
@@ -0,0 +1,290 @@
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def control(self):
+        return self._control
+
+    @property
+    def control_id(self):
+        return self._control_id
+
+    @property
+    def items(self):
+        return self._items
+
+    @property
+    def required(self):
+        if self._required:
+            return self._required
+        else:
+            return False
+
+
+class Component(object):
+    """
+    Represents a runtime-specific component_id
+    """
+
+    id: str
+    name: str
+    description: str
+    runtime: str
+    properties: List[ComponentProperty]
+    op: str
+
+    def __init__(self, id: str, name: str, description: Optional[str], runtime: Optional[str] = None,
+                 properties: Optional[List[ComponentProperty]] = None, op: Optional[str] = None):
+        """
+        :param id: Unique identifier for a component_id
+        :param name: The name of the component_id for display
+        :param description: The description of the component_id
+        :param runtime: The runtime of the component_id (e.g. KFP or Airflow)
+        :param properties: The set of properties for the component_id
+        :type properties: List[ComponentProperty]
+        :param op: The operation name of the component_id; used by generic components in rendering the palette
+        """
+
+        if not id:
+            raise ValueError("Invalid component_id: Missing field 'id'.")
+        if not name:
+            raise ValueError("Invalid component_id: Missing field 'name'.")
+
+        self._id = id
+        self._name = name
+        self._description = description
+        self._runtime = runtime
+        self._properties = properties
+        self._op = op
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def runtime(self):
+        return self._runtime
+
+    @property
+    def properties(self):
+        return self._properties
+
+    @property
+    def op(self):
+        if self._op:
+            return self._op
+        else:
+            return self._id
+
+
+class ComponentReader(LoggingConfigurable):
+    """
+    Abstract class to model component_entry readers that can read components from different locations
+    """
+    type: str = None
+
+    @property
+    def type(self) -> str:
+        return self.type
+
+    @abstractmethod
+    def read_component_definition(self, component_id: str, location: str) -> str:
+        raise NotImplementedError()
+
+
+class FilesystemComponentReader(ComponentReader):
+    """
+    Read a component_id definition from the local filesystem
+    """
+    type = 'filesystem'
+
+    def read_component_definition(self, component_id: str, location: str) -> str:
+        component_location = os.path.join(os.path.dirname(__file__), location)
```

Comment:
This will need to be coordinated with the change pending in #1854 that adds appropriate prefixing logic to the `location` value.